lib/llhttp/parser.js

```diff
@@ -0,0 +1,217 @@
+'use strict'
+
+/* global WebAssembly */
+
+const { resolve } = require('path')
+const { readFileSync } = require('fs')
+const constants = require('./constants')
+const { kMaxHeadersSize } = require('../core/symbols')
+const WASM_BUILD = resolve(__dirname, './llhttp.wasm')
+const bin = readFileSync(WASM_BUILD)
+const mod = new WebAssembly.Module(bin)
+
+const kOnMessageBegin = 0
+const kOnHeaders = 1
+const kOnHeadersComplete = 2
+const kOnBody = 3
+const kOnMessageComplete = 4
+
+const kPtr = Symbol('kPrt')
+const kStatusMessage = Symbol('kStatusMessage')
+const kHeadersFields = Symbol('kHeadersFields')
+const kHeadersValues = Symbol('kHeadersValues')
+const kBufferSize = Symbol('kBufferSize')
+const kHeadersTimeout = Symbol('kHeadersTimeout')
+const kBufferPtr = Symbol('kBufferPtr')
+const kMallocBuffer = Symbol('kMallocBuffer')
+const kGetHeaders = Symbol('kGetHeaders')
+const kReset = Symbol('kReset')
+const kMakeError = Symbol('kMakeError')
+
+const cstr = (ptr, len) =>
+  Buffer.from(inst.exports.memory.buffer, ptr, len).toString()
+
+/* eslint-disable camelcase */
+const wasm_on_message_begin = p => {
+  const i = instMap.get(p)
+  i[kReset]()
+  return i[kOnMessageBegin]()
+}
+
+const wasm_on_url = (p, at, length) => {
+  return 0
+}
+
+const wasm_on_status = (p, at, length) => {
+  instMap.get(p)[kStatusMessage] = cstr(at, length)
+  return 0
+}
+
+const wasm_on_header_field = (p, at, length) => {
+  const i = instMap.get(p)
+  i[kHeadersFields].push(cstr(at, length))
+  return 0
+}
+
+const wasm_on_header_value = (p, at, length) => {
+  const i = instMap.get(p)
+  i[kHeadersValues].push(cstr(at, length))
+  return 0
+}
+
+const wasm_on_headers_complete = p => {
+  const i = instMap.get(p)
+  const versionMajor = inst.exports.llhttp_get_http_major(p)
+  const versionMinor = inst.exports.llhttp_get_http_minor(p)
+  const rawHeaders = i[kGetHeaders]()
+  const statusCode = inst.exports.llhttp_get_status_code(p)
+  const statusMessage = i[kStatusMessage]
+  const upgrade = Boolean(inst.exports.llhttp_get_upgrade(p))
+  const shouldKeepAlive = inst.exports.llhttp_should_keep_alive(p)
+
+  return i[kOnHeadersComplete](versionMajor, versionMinor, rawHeaders, null,
+    null, statusCode, statusMessage, upgrade, shouldKeepAlive)
+}
+
+const wasm_on_body = (p, at, length) => {
+  const i = instMap.get(p)
+  const u8 = new Uint8Array(inst.exports.memory.buffer, at, length)
+  const body = Buffer.from(u8) // llhttp re-uses buffer so we need to make a copy.
```

Comment:
We could also make this the responsibility of the user, i.e. they must consume buffer synchronously. Move this logic to all api methods expect for `dispatch` and add a note in dispatch about this.

Reply:
Maybe future optimization.

Reply:
Maybe add a TODO?

Reply:
I'll add a TODO 👍🏻